extensions/stackdriver/stackdriver.cc

```diff
@@ -0,0 +1,96 @@
+/* Copyright 2019 Istio Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// NOLINT(namespace-envoy)
+#include <google/protobuf/util/json_util.h>
+#include <random>
+#include <string>
+#include <unordered_map>
+
+#include "stackdriver.h"
+
+#ifndef NULL_PLUGIN
+#include "api/wasm/cpp/proxy_wasm_intrinsics.h"
```

Comment:
Can you explain why the imports are different? Why does it matter which namespace this code lives in?

Reply:
The header file needs to be different, because native plugin API has different underlying implementation and has to include a different header file. 

I think the native runtime expects each plugin to be implemented in a sub namespace of `Envoy::Extensions::Common::Wasm::Null::Plugin`. For instance, the [plugin registry factory](https://github.com/envoyproxy/envoy-wasm/blob/b21cda033caae515a9f2d07de8cd369ffbd01051/source/extensions/common/wasm/null/null.h#L123) is expected to be included in such a namespace to make sure the code is the same with different runtimes. Alternative is to replace namespace block with `using Envoy::Extensions::Common::Wasm::Null::Plugin`, but it won't successfully compile due to that registry macro and I guess it needs some work in envoy-wasm code.

Reply:
One way to work around is to replace namespace block with two using declaration:
```
diff --git a/extensions/stackdriver/stackdriver.h b/extensions/stackdriver/stackdriver.h
index d0d3936..b1f60ef 100644
--- a/extensions/stackdriver/stackdriver.h
+++ b/extensions/stackdriver/stackdriver.h
@@ -21,15 +21,9 @@
 #ifndef NULL_PLUGIN
 #include "api/wasm/cpp/proxy_wasm_intrinsics.h"
 #else
-
 #include "extensions/common/wasm/null/null.h"
-
-namespace Envoy {
-namespace Extensions {
-namespace Common {
-namespace Wasm {
-namespace Null {
-namespace Plugin {
+using namespace Envoy::Extensions::Common::Wasm::Null;
+using namespace Envoy::Extensions::Common::Wasm::Null::Plugin;
 #endif
 
 namespace Stackdriver {
@@ -70,13 +64,5 @@ static RegisterContextFactory register_StackdriverContext(
 
 }  // namespace Stackdriver
 
-#ifdef NULL_PLUGIN
-}  // namespace Plugin
-}  // namespace Null
-}  // namespace Wasm
-}  // namespace Common
-}  // namespace Extensions
-}  // namespace Envoy
-#endif

 ```
And I think it is not quite ideal to include `using` in header file, so the original way might still be preferred I think.

Reply:
I believe this is needed right now, not exactly sure why, though.

@jplevyak can tell us next week.